core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -36,12 +38,33 @@
  */
 public class CurrentIteration implements ValueExpression {
 
+    private final ValueExpression level;
+
+    public CurrentIteration(final ValueExpression level) {
+        this.level = level;
+    }
+
     @Override
     public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
-        if (parseState.iterations.head == null) {
+        int level = getLevel(parseState, encoding);
+        ImmutableList<BigInteger> iterations = parseState.iterations;
+        while (level > 0 && iterations.tail != null) {
```

Comment:
I suspect this will not pass @jvdb's style checker 😁 Reassigning two local variables... You can check up-front if the `level` value is not too large compared to the size of `iterations`, eliminating the need to decrease it inside a loop. Finding the correct iteration can be achieved using a recursive method (which I think would be safe enough, unless we parse *really* deep data structures).

Reply:
Ok. Using an recursive method instead.